espnet2/bin/asr_transducer_inference.py

```diff
@@ -0,0 +1,779 @@
+
+    """
+    assert check_argument_types()
+
+    if batch_size > 1:
+        raise NotImplementedError("batch decoding is not implemented")
+    if ngpu > 1:
+        raise NotImplementedError("only single GPU decoding is supported")
+
+    logging.basicConfig(
+        level=log_level,
+        format="%(asctime)s (%(module)s:%(lineno)d) %(levelname)s: %(message)s",
+    )
+
+    if ngpu >= 1:
+        device = "cuda"
+    else:
+        device = "cpu"
+
+    # 1. Set random-seed
+    set_all_random_seed(seed)
+
+    # 2. Build speech2text
+    speech2text_kwargs = dict(
+        asr_train_config=asr_train_config,
+        asr_model_file=asr_model_file,
+        beam_search_config=beam_search_config,
+        lm_train_config=lm_train_config,
+        lm_file=lm_file,
+        token_type=token_type,
+        bpemodel=bpemodel,
+        device=device,
+        dtype=dtype,
+        beam_size=beam_size,
+        lm_weight=lm_weight,
+        nbest=nbest,
+        quantize_asr_model=quantize_asr_model,
+        quantize_modules=quantize_modules,
+        quantize_dtype=quantize_dtype,
+        streaming=streaming,
+        chunk_size=chunk_size,
+        left_context=left_context,
+        right_context=right_context,
+    )
+    speech2text = Speech2Text.from_pretrained(
+        model_tag=model_tag,
+        **speech2text_kwargs,
+    )
+
+    # 3. Build data-iterator
+    loader = ASRTransducerTask.build_streaming_iterator(
+        data_path_and_name_and_type,
+        dtype=dtype,
+        batch_size=batch_size,
+        key_file=key_file,
+        num_workers=num_workers,
+        preprocess_fn=ASRTransducerTask.build_preprocess_fn(
+            speech2text.asr_train_args, False
+        ),
+        collate_fn=ASRTransducerTask.build_collate_fn(
+            speech2text.asr_train_args, False
+        ),
+        allow_variable_data_keys=allow_variable_data_keys,
+        inference=True,
+    )
+
+    # 4 .Start for-loop
+    with DatadirWriter(output_dir) as writer:
+        for keys, batch in loader:
+            assert isinstance(batch, dict), type(batch)
+            assert all(isinstance(s, str) for s in keys), keys
+
+            _bs = len(next(iter(batch.values())))
+            assert len(keys) == _bs, f"{len(keys)} != {_bs}"
+            batch = {k: v[0] for k, v in batch.items() if not k.endswith("_lengths")}
+            assert len(batch.keys()) == 1
+
+            try:
+                if streaming and chunk_size > 0:
+                    speech = batch["speech"]
+
+                    _steps = len(speech) // speech2text._raw_ctx
+                    _end = 0
+
+                    for i in range(_steps):
+                        _end = (i + 1) * speech2text._raw_ctx
+
+                        speech2text(
+                            speech[i * speech2text._raw_ctx : _end], is_final=False
+                        )
+
+                    final_hyps = speech2text(speech[_end : len(speech)], is_final=True)
```

Comment:
I don't mind, it would just mean splitting `speech2text.__call___`!
Currently, if `streaming: False` (default), batch of inputs will be passed and the following will be executed: 

```python
            self.asr_model.encoder.dynamic_chunk_training = False

            enc_out, _ = self.asr_model.encoder(feats, feats_length)

            nbest_hyps = self.beam_search(enc_out[0])
```

Same as offline, `foward(...)` methods will be called instead of `chunk_forward(...)`

Reply:
For non streaming I used `__call__` and added a `streaming_decode(...)` method! Is that OK like this?

Reply:
Sure! I think it is ok.